doc/oper.xml

```diff
@@ -1300,6 +1300,40 @@ false
 </ManSection>
 <#/GAPDoc>
 
+<#GAPDoc Label="VerticesReachableFrom">
+<ManSection>
+  <Oper Name="VerticesReachableFrom" Arg="digraph, root"/>
+  <Returns>A list.</Returns>
+  <Description>
+    This operation returns a list of the vertices  <A>v</A> , for which there exists
+    a non-trivial directed walk from vertex <A>root</A> to vertex <A>v</A> in the digraph
```

Comment:
That's not quite true - the function currently always returns the root to be reachable from the root, even if it is only reachable by a *trivial* directed walk, e.g.:
```
gap> D := Digraph([[2], []]);;
gap> VerticesReachableFrom(D, 1);
[ 1, 2 ]
```
I think the currently-documented behaviour is the right behaviour, so please update the code to match it 🙂

Reply:
Thank you @wilfwilson ! I've fixed this, but for some reason if I try to just remove the root vertex from the list of visited vertices and start off with an empty list instead of a list containing  the root the function does not return output (probably runs forever for some reason). I had this problem in the beginning and that is basically why I included the root. Still haven't found what I'm missing. I removed the root at the end when we have performed the DFS, but I guess this is not the best way to do it :)

Reply:
Yeah it might be fiddly! The problem with removing the root at the end is that that's not always a valid thing to do. Indeed the root is quite often reachable from itself via a *non-trivial* walk. For instance, with `D := CompleteDigraph(2)`, then `VerticesReachableFrom(D, 1)` should be `[ 1, 2 ]` because `1 → 2 → 1` and `1 → 2` are non-trivial walks from 1 to 1 and 1 to 2 respectively.

Reply:
oh yes, right! That doesn't solve the problem! :(
I'll try to find some other way of doing this, but for the time being I can't really see what causes this.

Reply:
@wilfwilson my thinking might be completely wrong, but since we execute a depth first search to find which vertices are reachable from a root vertex, isn't the only case we meet the root vertex the start of the process, despite other possible non-trivial paths? Also I did a google search and reachability is mentioned to be reflexive, due to the trivial path of zero length, so maybe I need to change documentation instead, or if we're also interested in non-trivial cases make a second function? :)

Reply:
I understand what you're saying. 🙂

I'm sure there are some sources (perhaps even most?) that consider each vertex to be reachable from itself. Mathematically, that is probably the cleanest definition. But when writing a software package, as well as the purity of the mathematics, we have to think about:

1. what is the most *useful* thing to do? and
2. does our package behave consistently?

To the first point, it's surely the most useful to have the definition with a non-trivial walk, since that gives you *extra* information. If a user prefers this definition, then our function works perfectly for them; if a user prefers the definition with an arbitrary walk, then our function can work fine for them - they just have to remember to add the root itself to the result if it is not already in there. But you can't go the other way: if `VerticesReachableFrom(D, v)` *always* includes `v`, then there is no way to recover the information about whether or not `v` is reachable from itself by a non-trivial walk! So the function is not very useful to someone who only wants to know about non-trivial walks.

For the second point, we already have a function `IsReachable(D, u, v)` which (according to its documentation) only returns `true` if there's a non-trivial walk in `D` from `u` to `v:
```
gap> D := ChainDigraph(5);; IsReachable(D, 1, 1);
false
gap> D := CompleteDigraph(5);; IsReachable(D, 1, 1);
true
```
So we should remain consistent with the behaviour of this function, which already exists in previously published versions of the package.

> since we execute a depth first search to find which vertices are reachable from a root vertex, isn't the only case we meet the root vertex the start of the process, despite other possible non-trivial paths

No. In a depth-first search from a root `v`, we follow every edge that leaves any vertex that's reachable from `v` (including `v` itself). (By definition, these edges all point towards vertices that are also reachable from `v`). When we follow an edge, either it leads to a new vertex, which is great, otherwise it leads to a vertex that we've seen before - in which case we immediately backtrack. So we typically meet each vertex (including the root vertex, if it is reachable by a non-trivial walk) several times during the course of the search. It's just that we backtrack immediately for all but the first meeting.

I will comment on the code about how you can fix the code.

Reply:
Thank you so much @wilfwilson that makes absolute sense! :)